UCP/CORE: Implement flush+destroy for UCT EPs on UCP Worker

dmitrygx opened the pull request:
## What

Implement flush+destroy for UCT EPs on UCP Worker.

## Why ?

To be able to do `flush(CANCEL/LOCAL)` and then destroy UCT EP when UCP EP could be destroyed.
So, UCT EP will be destroyed on UCP Worker.
It fixes possible undone outstanding operation on some transports (e.g. there was an issue on RC).

## How ?

Implement `ucp_worker_discard_uct_ep()` that's called instead of `uct_ep_destroy()`

evgeny-leksikov commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,109 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    req->send.state.uct_comp.func  = ucp_worker_discard_uct_ep_flush_comp;
+    req->send.state.uct_comp.count = 1;
+
+    for (;;) {
```
how does it make sense without calling progress? I think need to reschedule `ucp_worker_discard_uct_ep_progress` for next try.
also pls make code more explicit, e.g.:
```
if (status == UCS_UCS_ERR_BUSY) {
    /* try again code */
}
```

dmitrygx replied:
if `uct_ep_pending_add()` completest with `UCS_UCS_ERR_BUSY`, then `uct_ep_flush()` will return either `UCS_INPROGRESS` or `UCS_OK`, that;s why we do `for(;;)` to make `uct_ep_flush()` again

evgeny-leksikov replied:
you mean UCS_NO_RESOURCES from flush and ERR_BUSY from pending_add are mutually exclusive cases?

dmitrygx replied:
> you mean UCS_NO_RESOURCES from flush and ERR_BUSY from pending_add are mutually exclusive cases?

yes, if `uct_ep_flush()` returns `UCS_ERR_NO_RESOURCES`, and then `uct_ep_pending_add()` returns `UCS_UCS_ERR_BUSY`, we could expect then a subsequent `uct_ep_flush()` call should be completed with `UCS_OK`/`UCS_INPROGRESS`

evgeny-leksikov replied:
how is it possible without [async]progress?

dmitrygx replied:
> how is it possible without [async]progress?

yes, this is possible only in case of async-progress.
in other cases, adding to pending queue should return success/in-progress after NO_RESOURCES from `uct_ep_flush()`

evgeny-leksikov replied:
pending add can't return IN_PROGRESS :)
IMHO better to keep code simpler for non-fast-path, also this case might happen very rare...

dmitrygx replied:
> pending add can't return IN_PROGRESS :)

yes, I meant `UCS_OK`
@evgeny-leksikov do you mean to re-register progress cb on UCT worker's callback queue and then retry?
@brminich @yosefe wdyt?

evgeny-leksikov replied:
BTW, rescheduling one-shot-GC-progress in resource critical state we can keep resources free for something eager

evgeny-leksikov replied:
> @evgeny-leksikov do you mean to re-register progress cb on UCT worker's callback queue and then retry?

yes, try flush/pending_add one time per CB invoke

dmitrygx replied:
> BTW, rescheduling one-shot-GC-progress in resource critical state we can keep resources free for something eager

yes, make sense to me

evgeny-leksikov commented on src/ucp/core/ucp_request.h:
```diff
@@ -209,8 +209,10 @@ struct ucp_request {
                 struct {
-                    ucp_worker_h          ucp_worker;     /* UCP worker where discard UCT EP operation
-                                                           * submitted on */
+                    uct_worker_cb_id_t    cb_id;          /* UCT worker callback ID to schedule
```
IMO better to use one-shot callback, this way no need to store id and light weight on "likely" path

dmitrygx replied:
done

evgeny-leksikov commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2395,10 +2407,11 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
+            goto try_again;
         }
     } else {
         if (status != UCS_INPROGRESS) {
```
minor: else if(...)

dmitrygx replied:
done

evgeny-leksikov commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2395,10 +2407,11 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
         if (status == UCS_ERR_BUSY) {
-            /* try again to flush UCT EP */
-            return 0;
+            goto try_again;
```
minor: can avoid `goto` with the same code indentation checking against UCS_OK:
```
status = uct_ep_pending_add(uct_ep, &req->send.uct, 0);
if (status == UCS_OK) {
    return 1;
}

ucs_assert(status == UCS_ERR_BUSY);
/* try again to flush UCT EP */
uct_worker_progress_register_safe(worker->uct,
                                  ucp_worker_discard_uct_ep_progress,
                                  req, UCS_CALLBACKQ_FLAG_ONESHOT,
                                  &cb_id);
return 0;
```

dmitrygx replied:
removed `goto` operator, but I prefer to handle `UCS_ERR_BUSY` first and have a single return point for success/in-progress (from flush)/errors

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,130 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    ucs_trace_req("flush completion req=%p status=%d", req, status);
+
+    /* don't destroy UCT EP from the flush completion callback,
+     * schedule a progress callback to destroy UCT EP */
```
i'dd add
schedule a progress callback "on the main thread"

dmitrygx replied:
done

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,130 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+
+    req = ucp_request_get(worker);
+    if (ucs_unlikely(req == NULL)) {
+        ucs_error("unable to allocate reqquest");
```
`request`, also can be a bit more verbose

dmitrygx replied:
done

dmitrygx commented:
bot:pipe:retest

dmitrygx commented:
/azp run

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2462,7 +2459,8 @@ void ucp_worker_discard_uct_ep(ucp_worker_h worker, uct_ep_h uct_ep,
     if (ucs_unlikely(req == NULL)) {
-        ucs_error("unable to allocate reqquest");
+        ucs_error("unable to allocate request for discarding UCT EP %p "
+                  "on UCP workjer %p", uct_ep, worker);
```
`worker`

dmitrygx replied:
done

evgeny-leksikov commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,128 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+                                              req, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                              &cb_id);
+            return 0;
+        } else if (status != UCS_OK) {
```
better to assert "impossible" flow, `uct_ep_pending_add` can't return other error by its doc

dmitrygx replied:
done

dmitrygx commented:
@yosefe could you review pls?

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,125 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+        /* destroy WIREUP EP allocated for this UCT EP, since
+         * discard operation most likely won't have an access to
+         * UCP EP as it could be destroyed by the caller */
+        uct_ep_destroy(&wireup_ep->super.super);
```
what about aux_ep which could be inside wireup_ep?

dmitrygx replied:
done

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,125 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+        return;
+    }
+
+    ++worker->flush_ops_count;
```
should be in UCS_ASYNC_BLOCK because wireup process can access it from async context

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,125 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
+    ucs_status_t status;
+
+    req->send.state.uct_comp.func  = ucp_worker_discard_uct_ep_flush_comp;
+    req->send.state.uct_comp.count = 1;
+
+    status = uct_ep_flush(uct_ep, req->send.discard_uct_ep.ep_flush_flags,
+                          &req->send.state.uct_comp);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        req->send.uct.func = ucp_worker_discard_uct_ep_pending_cb;
```
we can initialize these request fields in ucp_worker_discard_uct_ep (then no need for forward declaration)

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,125 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+                                                send.state.uct_comp);
+    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
+
+    ucs_trace_req("flush completion req=%p status=%d", req, status);
```
`ucp_trace_req(req, "discard_ep flush completion status %s", ucs_status_string(status));`

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,125 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    uct_ep_h uct_ep     = req->send.discard_uct_ep.uct_ep;
+    ucp_worker_h worker = req->send.discard_uct_ep.ucp_worker;
+
+    ucs_debug("destroy uct_ep=%p", uct_ep);
```
ucp_trace_req

yosefe reviewed (commented):
is it possible to add gtest?

dmitrygx commented:
> is it possible to add gtest?

added the tests

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2443,6 +2440,13 @@ void ucp_worker_discard_uct_ep(ucp_worker_h worker, uct_ep_h uct_ep,
+            /* discard the WIREUP EP's auxiliary EP */
+            ucp_worker_discard_uct_ep(worker, wireup_ep->aux_ep,
+                                      ep_flush_flags);
+            wireup_ep->aux_ep = NULL;
```
ucp_wireup_ep_disown(uct_ep, aux_ep)

dmitrygx replied:
done

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,130 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
+    ucs_status_t status;
+
+    req->send.state.uct_comp.func  = ucp_worker_discard_uct_ep_flush_comp;
+    req->send.state.uct_comp.count = 1;
```
maybe set this in ucp_worker_discard_uct_ep?

dmitrygx replied:
ok, done

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,130 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    return ucp_worker_discard_uct_ep_progress(req) ? UCS_OK : UCS_INPROGRESS;
+}
+
+/* must be called with async lock held */
```
put this in H file

dmitrygx replied:
done

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2350,130 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    ucp_trace_req(req, "destroy uct_ep=%p", uct_ep);
+    uct_ep_destroy(uct_ep);
+    ucp_request_put(req);
+    --worker->flush_ops_count;
```
need UCS_ASYNC_BLOCK here?

dmitrygx replied:
yes, good catch
done

yosefe commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,223 @@
+        }
+
+        do {
+            progress();
+
+            if (!m_flush_comps.empty()) {
+                uct_completion_t *comp = m_flush_comps.back();
+
+                m_flush_comps.pop_back();
+                uct_invoke_completion(comp, UCS_OK);
+            }
+
+            if (!m_pending_reqs.empty()) {
+                uct_pending_req_t *req = m_pending_reqs.back();
+
+                ucs_status_t status = req->func(req);
+                if (status == UCS_OK) {
+                    m_pending_reqs.pop_back();
+                } else {
+                    EXPECT_EQ(UCS_INPROGRESS, status);
+                }
+            }
+        } while (m_destroyed_ep_count < m_created_ep_count);
```
can't we just call ucp_worker_flush() and afterwards expect all eps were discarded?

dmitrygx replied:
but we still need to call flush_completion or peding_req's cb
so, this loop is needed anyway

yosefe replied:
so maybe do nonblocking flush and call these in the wait loop?
we want to test that when worker is flushed, all eps are destroyed..

dmitrygx replied:
> so maybe do nonblocking flush and call these in the wait loop?

yes, good point
done

> we want to test that when worker is flushed, all eps are destroyed..

yes, I didn't think about this

brminich commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,224 @@
+}
+
+UCS_TEST_P(test_ucp_worker_discard, wireup_ep_flush_in_progress) {
+    test_worker_discard((void*)ucs_empty_function_return_success,
```
why return success?

dmitrygx replied:
fixed

brminich commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,224 @@
+}
+
+UCS_TEST_P(test_ucp_worker_discard, wireup_ep_flush_no_resource_pending_add_busy) {
+    test_worker_discard((void*)ucs_empty_function_return_success,
```
why return success?

dmitrygx replied:
fixed

brminich commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,224 @@
+}
+
+UCS_TEST_P(test_ucp_worker_discard, wireup_ep_flush_no_resource_pending_add_ok_then_busy) {
+    test_worker_discard((void*)ucs_empty_function_return_success,
```
why return success?

dmitrygx replied:
fixed

yosefe commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -92,6 +93,10 @@ class test_ucp_worker_discard : public ucp_test {
         }
 
+        void *flush_nb_req = sender().flush_worker_nb(0);
+        EXPECT_TRUE((flush_nb_req != NULL) && !UCS_PTR_IS_ERR(flush_nb_req) &&
+                    (ucp_request_check_status(flush_nb_req) == UCS_INPROGRESS));
```
break it to multiple EXPECT_xx

dmitrygx replied:
no need anymore

yosefe commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -114,6 +119,12 @@ class test_ucp_worker_discard : public ucp_test {
             }
         } while (m_destroyed_ep_count < m_created_ep_count);
 
+        if (UCS_PTR_IS_PTR(flush_nb_req)) {
+            status = ucp_request_check_status(flush_nb_req);
+            EXPECT_UCS_OK(status);
+            ucp_request_release(flush_nb_req);
+        }
```
we should check flush completion in the do/while loop
and after loop check all eps are destroyed

dmitrygx replied:
done, checking `ucp_worker_t::flush_ops_count` field as discussed offline

yosefe commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -119,12 +116,7 @@ class test_ucp_worker_discard : public ucp_test {
-            ucp_request_release(flush_nb_req);
-        }
-
+        EXPECT_EQ(0, sender().worker()->flush_ops_count);
```
do {
} while (flush_ops_count > 0)

EXPECT_EQ(created_ep_count, destroyed_ep_count)

dmitrygx replied:
fixed

yosefe commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,224 @@
+            ucp_worker_discard_uct_ep(sender().worker(), discard_ep,
+                                      UCT_FLUSH_FLAG_LOCAL);
+        }
+        do {
```
space before

dmitrygx replied:
done

dmitrygx commented:
bot:pipe:retest

dmitrygx commented:
@evgeny-leksikov could you review pls?
I've changed the implementation where we changed the order to check TX window emptiness before checking the ability to do TX in EP flush. This breaks some tests: data corruptions for the tests that check the data that was received (or places using RMA primitives). It happens, seems we reported UCS_OK from EP flush when detected that TX window is empty now, but UD isn't really able to send data.
So, the fix proposed in the last changes in this PR:
reset `tx::max_psn` to `tx::psn + ca:cwnd` in `uct_ep_flush(CANCEL)` - so it means that we are able to send, and `uct_iface_flush()`/`uct_ep_flush(LOCAL)` will return UCS_OK.
Before this fix, we reported `UCS_NO_RESOURCE` from `uct_iface_flush()`/`uct_ep_flush(LOCAL), since `tx::max_psn` == `tx::psn`

dmitrygx commented:
bot:pipe:retest

dmitrygx commented:
/azp run

dmitrygx commented on src/uct/ib/ud/base/ud_ep.c:
```diff
@@ -221,7 +225,7 @@ static void uct_ud_ep_purge_outstanding(uct_ud_ep_t *ep)
 static void uct_ud_ep_purge(uct_ud_ep_t *ep, ucs_status_t status)
 {
-    uct_ud_ep_tx_stop(ep);
+    uct_ud_ep_reset_max_psn(ep);
```
it fixes doing `uct_iface_flush()`/`uct_ep_flush(LOCAL)` after `uct_ep_flush(CANCEL)` was done
before this fix, `uct_iface_flush()`/`uct_ep_flush(LOCAL)` returned `UCS_ERR_NO_RESOURCE` after canceling, since `no_window()` was true (it checks that `tx.psn == tx.max_psn`)

added a comment to clarify this fact

dmitrygx commented:
@brminich @evgeny-leksikov could you check if this PR looks good to you now?

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2373,153 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    return ucp_worker_discard_uct_ep_progress(req) ? UCS_OK : UCS_INPROGRESS;
```
if flush returned NO_RESOURCE, need to return NO_RESOURCE from this function as well, to desched the ep

dmitrygx replied:
done

yosefe commented on src/ucp/rma/flush.c:
```diff
@@ -444,16 +444,17 @@ static unsigned ucp_worker_flush_progress(void *arg)
 
     status = ucp_worker_flush_check(worker);
-    if ((status == UCS_OK) || (&next_ep->ep_list == &worker->all_eps)) {
+    if ((status == UCS_OK) && (&next_ep->ep_list == &worker->all_eps)) {
```
isn't this disabling endpoint-by-endpoint flush? if we always require iface flush to be completed?
maybe we can check just flush_ops_count?

dmitrygx replied:
> isn't this disabling endpoint-by-endpoint flush?

sorry, do you mean that it could disable case when we all `uct_ep_flush()`s are completed with `UCS_OK`, but `uct_iface_flush()` returns `UCS_ERR_NO_RESOURCE`?

dmitrygx commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2373,153 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    kh_del(ucp_worker_discard_uct_ep_hash,
+           &worker->discard_uct_ep_hash, iter);
+    uct_ep_destroy(uct_ep);
+    UCS_ASYNC_UNBLOCK(&worker->async);
```
makes sense to me
fixed

dmitrygx commented:
@evgeny-leksikov @brminich could you review pls?
purging was implemented, the two last commits add it

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2376,179 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    ucp_trace_req(req, "discard_uct_ep flush completion status %s",
+                  ucs_status_string(status));
+
+    /* don't destroy UCT EP from the flush completion callback, schedule
```
👍

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2350,3 +2376,179 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+            if (is_owner /* purge from the next UCT EP, if the WIREUP EP is
+                          * an owner for it */) {
+                /* purge the next UCT EP first to pass WIREUP MSGs to the user
+                 * purge callback, otherwise, they will be release when purging
```
minor: `released`

dmitrygx replied:
fixed

brminich commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,316 @@
+            m_created_ep_count++;
+
+            unsigned expected_pending_purge_reqs_count =
+                (ep_pending_purge_func == ep_pending_purge_func_iter) ?
```
minor: quite hard to read, i'd use `if/else` and add general comment (not for every line)

dmitrygx replied:
fixed

brminich commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -77,12 +77,25 @@ class test_ucp_worker_discard : public ucp_test {
+
+            if (ep_pending_purge_func == ep_pending_purge_func_iter) {
+                /* expected purging count is the number of pending
+                * reqyests in the UCT EP, in the WIREUP EP (if used)
```
requests

dmitrygx replied:
fixed

dmitrygx commented:
@evgeny-leksikov @yosefe could you review pls?

evgeny-leksikov commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,364 @@
+
+
+UCS_TEST_P(test_ucp_worker_discard, flush_ok) {
+    test_worker_discard((void*)ucs_empty_function_return_success,
+                        (void*)ucs_empty_function_do_assert,
+                        (void*)ucs_empty_function);
```
pls add arg names in comments

evgeny-leksikov reviewed (commented):
just few minor comments

dmitrygx commented:
@brminich @evgeny-leksikov @yosefe could you review pls?
IO demo failure is unrelated

evgeny-leksikov commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -278,22 +309,28 @@ class test_ucp_worker_discard : public ucp_test {
+std::vector<uct_completion_t*>  test_ucp_worker_discard::m_flush_comps;
 std::vector<uct_pending_req_t*> test_ucp_worker_discard::m_pending_reqs;
-std::map<uct_ep_h, unsigned> test_ucp_worker_discard::m_pending_reqs_count;
+std::map<uct_ep_h,
+         std::vector<uct_pending_req_t*> >
```
minor: can define type by `typedef` to make it shorter

dmitrygx replied:
done

yosefe commented:
/azp run

yosefe commented on src/ucp/wireup/wireup_ep.c:
```diff
@@ -219,30 +235,78 @@ static ucs_status_t ucp_wireup_ep_pending_add(uct_ep_h uct_ep,
+}
+
+void
+ucp_wireup_ep_pending_purge_common(ucp_wireup_ep_t *wireup_ep,
+                                   uct_pending_purge_callback_t wireup_msg_cb,
+                                   void *wireup_msg_arg,
+                                   uct_pending_purge_callback_t user_msg_cb,
+                                   void *user_msg_arg)
```
this is over complication
why not mostly keep the old flow and just pass flag whether to release wireup messages or call the purge_cb for them

dmitrygx replied:
we have to set wireup_ep to the `NULL` for debug purposes (to not touch it in the WIREUP MSG pending progress function) and to check that wireup_ep is `NULL` and to not decrement the pending count in this case

dmitrygx commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2390,190 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+
+    ucs_assert(wireup_ep != NULL);
+
+    if (purge_cb != NULL) {
```
removed

dmitrygx commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2386,180 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+
+    ucs_assert(uct_ep != NULL);
+
+    if (ucp_wireup_ep_test(uct_ep)) {
```
done

dmitrygx commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2390,190 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+         * the callback to not invoke it several times */
+        if (status_add == UCS_ERR_BUSY) {
+            uct_worker_progress_register_safe(worker->uct,
+                                              ucp_worker_discard_uct_ep_destroy_progress,
```
> 1\. it should be ucp_worker_discard_uct_ep_progress?

yes, this was a bug

> 2\. can we get rid of ucp_worker_discard_uct_ep_progress wrapper altogether? and in here just goto uct_pending_add again instead of scheduling progress

do you mean to do `uct_ep_flush()`+`uct_ep_pending_add()` again in case of `UCS_ERR_BUSY`?

dmitrygx replied:
I think, better to put it on progress, since it could be progressed and don't take too much time for flush+add

dmitrygx replied:
> > 1. it should be ucp_worker_discard_uct_ep_progress?
> 
> yes, this was a bug

added additional checks in the test code to verify it

dmitrygx commented:
@yosefe is it know issue with IO demo?
should I restart?

dmitrygx commented:
bot:pipe:retest

brminich commented on src/ucp/rma/flush.c:
```diff
@@ -444,17 +446,25 @@ static unsigned ucp_worker_flush_progress(void *arg)
-        /* Error returned from uct iface flush */
-        ucp_worker_flush_complete_one(req, status, 1);
-    } else if (worker->context->config.ext.flush_worker_eps) {
+    if (ucp_worker_flush_ops_count_check(worker)) {
```
why not leaving this check inside `ucp_worker_flush_check`? seems unnecessary
also checking  `!worker->flush_ops_count` looks more clear than `ucp_worker_flush_ops_count_check` name

dmitrygx replied:
fixed

brminich commented on src/ucp/wireup/wireup_ep.c:
```diff
@@ -22,6 +22,18 @@
+    ucp_wireup_ep_t              *wireup_ep;         /* WIREUP EP that owns the UCT
+                                                      * pending request for the WIREUP
+                                                      * MSG */
+    uct_pending_purge_callback_t wireup_msg_cb;      /* UCT pending purge cakkback that
```
callback

brminich commented on src/ucp/wireup/wireup_ep.c:
```diff
@@ -219,30 +235,78 @@ static ucs_status_t ucp_wireup_ep_pending_add(uct_ep_h uct_ep,
+    ucp_wireup_ep_pending_purge_arg_t *purge_arg =
+        (ucp_wireup_ep_pending_purge_arg_t*)arg;
+    ucp_wireup_ep_t *wireup_ep                   = purge_arg->wireup_ep;
+    uct_ep_h uct_ep                              =
```
minor: imo, using single line without alignment by `=` would be more readable in this case

brminich commented on src/ucp/wireup/wireup_ep.c:
```diff
@@ -219,30 +235,78 @@ static ucs_status_t ucp_wireup_ep_pending_add(uct_ep_h uct_ep,
+    /* do purging on AUX EP or on UCT EP if WIREUP is an owner of it */
+    if ((uct_ep == wireup_ep->aux_ep) || wireup_ep->super.is_owner) {
+        /* need to NULL the WIREUP EP in the WIREUP MSG proxy pending request
+         * to avoid dereferencing it when progressing prnding requests, since
```
pending

brminich commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -0,0 +1,462 @@
+            pending_reqs[base + i] = req;
+
+            if (func == ucp_wireup_msg_progress) {
+                
```
extra line

dmitrygx replied:
removed

dmitrygx commented:
bot:pipe:retest

dmitrygx commented:
@evgeny-leksikov @brminich could you review pls?

evgeny-leksikov commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -286,28 +339,20 @@ class test_ucp_worker_discard : public ucp_test {
 
-            std::vector<uct_pending_req_t*> *req_vec = &it->second;
+        for (unsigned i = 0; i < m_pending_purge_reqs_count; i++) {
+            std::vector<uct_pending_req_t*> *req_vec = &test_info->pending_reqs;
```
std::vector<uct_pending_req_t*> &req_vec = ...

dmitrygx replied:
done

evgeny-leksikov commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2440,33 +2440,25 @@ ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self,
+            /* need to remove from the pending queue */
+            status = UCS_OK;
+        } else {
+            ucs_assert(status != UCS_ERR_BUSY);
```
this assert looks wrong since status value is returned by `flush` here

dmitrygx replied:
this is not wrong, just to make sure that we will not enter the endless loop if `uct_ep_flush()` returns `UCS_ERR_BUSY`
added the comment

evgeny-leksikov replied:
but uct_ep_flush doc does not say that it cannot return `UCS_ERR_BUSY`

dmitrygx replied:
> but uct_ep_flush doc does not say that it cannot return `UCS_ERR_BUSY`

yes, will handle it

@yosefe what do you think to use progress_register() instead of this loop?
seems to be error-prone or overcomplicated if it will have several `retrun`/`break`/`continue` in the loop

evgeny-leksikov commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2440,33 +2440,25 @@ ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self,
+        } else if (status == UCS_INPROGRESS) {
+            /* need to remove from the pending queue */
+            status = UCS_OK;
+        } else {
```
UCS_OK is not handled?

dmitrygx replied:
it is handled in the `else` section

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2390,180 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    }
+
+    ++worker->flush_ops_count;
+    iter = kh_put(ucp_worker_discard_uct_ep_hash,
```
minor: maybe just `kh_put(ucp_worker_discard_uct_ep_hash,...`? Then no need in unused `iter` var

dmitrygx replied:
done

brminich commented on src/ucp/rma/flush.c:
```diff
@@ -384,10 +384,6 @@ static ucs_status_t ucp_worker_flush_check(ucp_worker_h worker)
     ucp_worker_iface_t *wiface;
     ucs_status_t status;
 
-    if (worker->flush_ops_count) {
```
why is removed here and moved outside?

dmitrygx replied:
we want to check for flush_ops count separately from `uct_iface_flush()` that we check in this function

brminich replied:
why, seems can just leave it returning UCS_INPROGRESS

dmitrygx replied:
> why, seems can just leave it returning UCS_INPROGRESS?

added a comment to make it clear
if it returns UCS_INPROGESS, but going over all EPs were done, we complete with `UCS_OK`, but we shouldn't

brminich commented on src/ucp/wireup/wireup_ep.c:
```diff
@@ -141,8 +141,8 @@ static uct_ep_h ucp_wireup_ep_get_msg_ep(ucp_wireup_ep_t *wireup_ep)
 {
-    ucp_request_t *proxy_req = ucs_container_of(self, ucp_request_t, send.uct);
-    uct_pending_req_t *req = proxy_req->send.proxy.req;
+    ucp_request_t *proxy_req   = ucs_container_of(self, ucp_request_t, send.uct);
```
single unrelated change in the file

dmitrygx replied:
fixed

dmitrygx commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -227,28 +247,61 @@ class test_ucp_worker_discard : public ucp_test {
         m_destroyed_ep_count++;
     }
 
+    static ep_test_info_t* ep_test_info_get(uct_ep_h ep) {
```
done

dmitrygx commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -227,28 +247,61 @@ class test_ucp_worker_discard : public ucp_test {
+        ep_test_info_map_t::iterator it = m_ep_test_info_map.find(ep);
+
+        if (it == m_ep_test_info_map.end()) {
+            ep_test_info_t test_info = {};
```
done

brminich commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2454,7 +2454,10 @@ ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self)
             /* need to remove from the pending queue */
             status = UCS_OK;
         } else {
+            /* make sure that uct_ep_flush() doenst return UCS_ERR_BUSY to not
```
does not

dmitrygx replied:
fixed

dmitrygx commented:
bot:pipe:retest

evgeny-leksikov commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -29,6 +29,12 @@ class test_ucp_worker_discard : public ucp_test {
         unsigned                           pending_add_count;
+
+        ep_test_info_t() {
+            pending_reqs.clear();
```
not needed, std::vector has its own constructor.

dmitrygx replied:
done

evgeny-leksikov commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -29,6 +29,12 @@ class test_ucp_worker_discard : public ucp_test {
+
+        ep_test_info_t() {
+            pending_reqs.clear();
+            flush_count       = 0;
```
minor: better to use initializer list
```
ep_test_info_t() : flush_count(0), pending_add_count(0) {
}
```

dmitrygx replied:
done

evgeny-leksikov commented on test/gtest/ucp/test_ucp_worker.cc:
```diff
@@ -220,34 +226,31 @@ class test_ucp_worker_discard : public ucp_test {
-        ep_test_info_t *test_info = ep_test_info_get(ep);
-        test_info->flush_count++;
-        return test_info->flush_count;
+        ep_test_info_t &test_info = ep_test_info_get(ep);
+        test_info.flush_count++;
+        return test_info.flush_count;
```
minor: can be shorter
return ++ep_test_info_get(ep).flush_count;

dmitrygx replied:
done

evgeny-leksikov commented on src/ucp/wireup/wireup_ep.c:
```diff
@@ -141,8 +141,8 @@ static uct_ep_h ucp_wireup_ep_get_msg_ep(ucp_wireup_ep_t *wireup_ep)
-    ucp_request_t *proxy_req   = ucs_container_of(self, ucp_request_t, send.uct);
-    uct_pending_req_t *req     = proxy_req->send.proxy.req;
+    ucp_request_t *proxy_req = ucs_container_of(self, ucp_request_t, send.uct);
+    uct_pending_req_t *req = proxy_req->send.proxy.req;
```
pls align on '='

dmitrygx replied:
it is not my code, I returned it back in the same state as it was

dmitrygx commented:
@evgeny-leksikov is it ok now?

dmitrygx commented:
@yosefe could you review pls?

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2390,185 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    status = uct_ep_flush(uct_ep, req->send.discard_uct_ep.ep_flush_flags,
+                          &req->send.state.uct_comp);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        status = uct_ep_pending_add(uct_ep, &req->send.uct, 0);
+        ucs_assert((status == UCS_ERR_BUSY) || (status == UCS_OK));
+        if (status == UCS_ERR_BUSY) {
+            uct_worker_progress_register_safe(worker->uct,
+                                              ucp_worker_discard_uct_ep_progress,
+                                              req, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                              &cb_id);
+        }
```
this is not completely right: we should do uct_ep_pending_add() only from ucp_worker_discard_uct_ep_progress() - the first time we are trying to execute the operation.
later on, when we are doing things from pending queue, should just return UCS_ERR_NO_RESOURCE, so the pending operation will remain on the pending queue.

dmitrygx replied:
fixed

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2390,185 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+
+    uct_ep_pending_purge(&wireup_ep->super.super, purge_cb, purge_arg);
+
+    if (wireup_ep->aux_ep != NULL) {
+        /* make sure that there is no WIREUP MSGs anymore */
+        uct_ep_pending_purge(wireup_ep->aux_ep,
```
shouldn't we do this before uct_ep_pending_purge()? or not at all?
ucp_wireup_ep_pending_purge already can purge aux_ep

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2390,185 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+            return;
+        }
+    } else if (purge_cb != NULL) {
+        uct_ep_pending_purge(uct_ep, purge_cb, purge_arg);
```
- uct_ep_pending_purge seems common for both cases
- no need to check purge_cb != NULL

yosefe commented on src/ucp/rma/flush.c:
```diff
@@ -444,17 +440,26 @@ static unsigned ucp_worker_flush_progress(void *arg)
-        /* Error returned from uct iface flush */
-        ucp_worker_flush_complete_one(req, status, 1);
-    } else if (worker->context->config.ext.flush_worker_eps) {
+    if (!worker->flush_ops_count) /* all scheduled operations on worker
+                                   * were completed */ {
```
weird placement of {

yosefe commented on src/ucp/rma/flush.c:
```diff
@@ -444,17 +440,26 @@ static unsigned ucp_worker_flush_progress(void *arg)
-        /* Error returned from uct iface flush */
-        ucp_worker_flush_complete_one(req, status, 1);
-    } else if (worker->context->config.ext.flush_worker_eps) {
+    if (!worker->flush_ops_count) /* all scheduled operations on worker
```
== 0

dmitrygx commented:
@yosefe could you review pls?

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2440,42 +2436,49 @@ ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self,
+        ucp_worker_discard_uct_ep_flush_comp(&req->send.state.uct_comp,
+                                             status);
+    }
+    return status;
```
if status != NO_RESOURCE should also return UCS_OK

yosefe replied:
if flush returned ok, who will call the completion?

dmitrygx replied:
> if status != NO_RESOURCE should also return UCS_OK

why? we should put to the pending and wait for the callback invocation



> if flush returned ok, who will call the completion?

it will be done in `ucp_worker_discard_uct_ep_flush_comp()`

yosefe replied:
- if the status is error which is not NO_RESOURCE, the flush failed, we cannot retry
- ok

yosefe commented on src/ucp/rma/flush.c:
```diff
@@ -444,17 +440,26 @@ static unsigned ucp_worker_flush_progress(void *arg)
-        /* Error returned from uct iface flush */
-        ucp_worker_flush_complete_one(req, status, 1);
-    } else if (worker->context->config.ext.flush_worker_eps) {
+    if (worker->flush_ops_count == 0) {
```
BTW, do we require iface_flush completion today before destroying UCP worker (AFAIK not)?
since it can leak some endpoints which are still being flushed

dmitrygx replied:
> BTW, do we require iface_flush completion today before destroying UCP worker (AFAIK not)?

we don't require, but now we require flushing all ops to be done and or iface_flush is completed or going over all EPs is done

yosefe replied:
need to check for leaks in worker destroy

dmitrygx replied:
done

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -2360,3 +2386,191 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    uct_ep_pending_purge(uct_ep, purge_cb, purge_arg);
+
+    if (ucp_wireup_ep_test(uct_ep)) {
+        uct_ep = ucp_worker_discard_wireup_ep(worker, ucp_wireup_ep(uct_ep),
+                                              ep_flush_flags,
+                                              purge_cb, purge_arg);
```
seems weird recursion: ucp_worker_discard_uct_ep->ucp_worker_discard_wireup_ep->ucp_worker_discard_uct_ep

maybe separate the 2nd part of this function to a helper:
- discard_uct_ep()
    if wireup:
      discard_wireup_ep()
      if next_ep_owner:
        discard_uct_ep_helper(next_ep) [works only on regular uct_ep, not wireup]
    else:
        discard_uct_ep_helper(uct_ep)

dmitrygx replied:
but what the problem with this recursion?
`ucp_worker_discard_uct_ep->ucp_worker_discard_wireup_ep` for WIREUP EP
and `->ucp_worker_discard_uct_ep` for AUX EP

yosefe replied:
imo would be better /wo recursion, since we know next_ep cannot be wireup_ep

dmitrygx replied:
done

dmitrygx commented:
@yosefe could you review pls?

changchengx commented:
Ported to https://github.com/yosefe/ucx/pull/223
